### ISSUE TITLE
"github.com/MontFerret/ferret/pkg/drivers" drivers.HTTPHeader json Marshal crashed

### ISSUE THREAD

**Issue by kaixinbuyu** [MontFerret/ferret#320]
**Question One:**

**Describe the bug**
"github.com/MontFerret/ferret/pkg/drivers"
 drivers.HTTPHeader json Marshal crashed

**To Reproduce**

```
func TestChromeHeader(t *testing.T) {
	res := make(drivers.HTTPHeader)

	j, err := json.Marshal(res)
	if err != nil {
		return
	}

	fmt.Printf(">>%s\n",j)
}

```
runtime: goroutine stack exceeds 1000000000-byte limit
fatal error: stack overflow

I have reviewed the code of HTTPHeader and found nothing wrong with HTTPHeader.String() func.

```

type HTTPHeader map[string][]string

func (h HTTPHeader) Type() core.Type {
	return HTTPHeaderType
}

func (h HTTPHeader) String() string {
	var buf bytes.Buffer

	for k := range h {
		buf.WriteString(fmt.Sprintf("%s=%s;", k, h.Get(k)))
	}

	return buf.String()
}

```

**Comment by kaixinbuyu** [comment]
And I find another question.

I currently skip the above error
by pass a raw header string and make a map[string]string to test SetExtraHTTPHeaders func 
but it seems not work . 
the SetExtraHTTPHeaders  func runs with no error **but  the cookie do not  take effect;**

```

//headerdata:I copyed from chrome's network-Request Headers in plain text(including Cookies)

headerdata = `
Cookie: web_uid=cEH0W1zdGlYYiFILB1UBAg==; UM_distinctid=16abfaeee3e117-04ba44469a782a-1b2a1209-1fa400-16abfaeee3fb37; _ga=GA1.2.257028262.1557994074; Z9M6_79fc_saltkey=i5XbqB95; Z9M6_79fc_lastvisit=1561789128; m_t_pinggu0518_m_t=1561792739327_1; _gid=GA1.2.629580353.1561792740; sso_hash=973e372c6ee5704ad4d0222b7b976425ef9c002844763d6a6d20fcccd29e39b1e75d624c3c4a19dd05b83da7be9b51018efb11e45eebd62b9ec53886a9f69fc22ef11dd7e9f717e91f3799ae40ca2cb9cf81366a31df7440ad888767f52a82e1cb1c8bd67bc3126b38d3ca282d8cb1f5a2c2320079fa46121579355694145cc57e5a8665791975943dc505db5ef6bb6f6a75ef959ce4a534b31482c9b9e8e004; sso_hash_low=02d93978317c3030ff3e137b81303fb9f01d69fd76c9c9f27c77ac7f90d0cadb24214efa4754f3687b3eaaeee1c2841869204ec9a626e7a029de327740fc1faa4e42abf0b189578e194c0d5bf85c1a48611058b0117bd213438eb6e91b86e8ebd9f545e2ca9db594cd3410d4ccbc30cc3929596e2718a30a890ef7cd92d2a05042b02c9d168716dc42adf8dc8e8f63f9265f61f17ebfd0b9a02708b3f4849500; dz_uid=136602; Z9M6_79fc_connect_is_bind=0; Hm_lvt_3a3fa94d4db32c2ae7bece34a6dae8ec=1561793155; Hm_lpvt_3a3fa94d4db32c2ae7bece34a6dae8ec=1561793155; MEIQIA_VISIT_ID=1NJJEloVYRwMqaqBZOpMFP2BzhU; Z9M6_79fc_ltzyts_zytk_136602=hide; Z9M6_79fc_sid=3xw0z5; Z9M6_79fc_ulastactivity=a7e7%2FtC4MfbXFX3mpeoU4kzNi5cM4mCIr1rbGGx%2FGmLKxm4xUIhC; Z9M6_79fc_sendmail=1; Z9M6_79fc_checkpm=1; Z9M6_79fc_auth=4dcb8IFxiW0vzhXpTlLttoZnBZGNH0a0ZX6PlvSRAqqraLVkqLPzWklTR4wyiC5GsPDesrdvgTIZY3%2BvlWy0Si1cDek; Z9M6_79fc_lastact=1561816456%09forum.php%09misc; CNZZDATA30069092=cnzz_eid%3D192894178-1557990224-%26ntime%3D1561812483
`


func (doc *HTMLDocument) SetHeaderData(ctx context.Context,headerdata values.String) error {
	res := make(map[string]string) // make(drivers.HTTPHeader)  //skip the above json.Marshal error 

	pairs := strings.Split(headerdata.String(), "\n") //\r\n
	for _, pair := range pairs {
		kv := strings.SplitN(pair, ":", 2)
		if len(kv) >= 2 {
			//res.Set(strings.Trim(kv[0], " "),strings.Trim(kv[1], " "))
			res[strings.Trim(kv[0], " ")] = strings.Trim(kv[1], " ")
		}
	}

	j, err := json.Marshal(res)
	if err != nil {
		return err
	}

	fmt.Printf(">>headers:%s\n",j)

	err = doc.client.Network.SetExtraHTTPHeaders(
		ctx,
		network.NewSetExtraHTTPHeadersArgs(network.Headers(j)),
	)

	if err != nil {
		return err
	}

	return nil
}

```

I have checked other language(nodejs,python) that the json passed to SetExtraHTTPHeaders is ok.I do't know what's wrong with my code.

**Comment by kaixinbuyu** [comment]
What I want to do is :
User Client Computer:
1.Open chrome.exe manually to open and login a url.
2.Copy Request Headers(with Cookies and other keys) in plain text  and save to database.

Server computer(Ferret):
3.Load headers string(from database) into ferret doc.client to check if login success.

**Comment by ziflex** [comment]
Hey @kaixinbuyu thank you for reporting this.
It seems it's a bug in ``MarshalJSON`` method of ``HTTPHeader``. It does not cast itself to ``map[string][]string`` before calling ``json.Marshal`` which triggers infinite loop.

**Comment by ziflex** [comment]
The fix has landed on master branch and will be a part of 0.8 release.
Feel free to re-open the issue if it still occurs.
